Add Check Support for Java 21 Pattern Matching for Switch Syntax: MagicNumber

child of #14961:
doc :https://checkstyle.org/checks/coding/magicnumber.html

---

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
               <module name="MagicNumber">

        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(boolean p, int x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
       if (obj instanceof Integer i && i < 3) { // violation
            System.out.println("null");
       } else {
            System.out.println("not null");
       }

       switch (obj) {
            case ColoredPoint(_,int x, int y) when (x < 3 && y != 6): // 2 violations
                System.out.println("null");
                break;
            default:
                System.out.println("not null");
       }
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:6:44: '3' is a magic number. [MagicNumber]
[ERROR] D:\CS\test\src\Test.java:13:57: '3' is a magic number. [MagicNumber]
[ERROR] D:\CS\test\src\Test.java:13:67: '6' is a magic number. [MagicNumber]
Audit done.
Checkstyle ends with 3 errors.
PS D:\CS\test> 
```
---
The check works correctly when the magic number is in guard. we may add an input file here.

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheckTest.java
@@ -727,4 +727,21 @@ public class MagicNumberCheckTest
         verifyWithInlineConfigParser(
                 getPath("InputMagicNumberMagicNumber3.java"), expected);
     }
+
+    @Test
+    public void testMagicNumberInGuards() throws Exception {
+        final String[] expected = {
+            "21:63: " + getCheckMessage(MSG_KEY, "3"),
+            "21:72: " + getCheckMessage(MSG_KEY, "8"),
+            "27:58: " + getCheckMessage(MSG_KEY, "3"),
+            "27:68: " + getCheckMessage(MSG_KEY, "6"),
+            "31:50: " + getCheckMessage(MSG_KEY, "10.88"),
+            "33:46: " + getCheckMessage(MSG_KEY, "6"),
+            "39:59: " + getCheckMessage(MSG_KEY, "0.5"),
+            "39:67: " + getCheckMessage(MSG_KEY, "5"),
+            "44:23: " + getCheckMessage(MSG_KEY, "6"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputMagicNumberMagicNumberInGuards.java"), expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberMagicNumberInGuards.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberMagicNumberInGuards.java
@@ -1,0 +1,51 @@
+/*
+MagicNumber
+ignoreNumbers = (default)-1, 0, 1, 2
+ignoreHashCodeMethod = (default)false
+ignoreAnnotation = (default)false
+ignoreFieldDeclaration = (default)false
+ignoreAnnotationElementDefaults = (default)true
+constantWaiverParentToken = (default)TYPECAST, METHOD_CALL, EXPR, ARRAY_INIT, UNARY_MINUS, \
+                            UNARY_PLUS, ELIST, STAR, ASSIGN, PLUS, MINUS, DIV, LITERAL_NEW
+tokens = (default)NUM_DOUBLE, NUM_FLOAT, NUM_INT, NUM_LONG
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.magicnumber;
+
+public class InputMagicNumberMagicNumberInGuards {
+
+    void m(Object o) {
+        if (o instanceof Point(int x, int y, double z) && x < 3 && y > 8) {}
+        // 2 violations above:
+        //                  ''3' is a magic number'
+       //                   ''8' is a magic number'
+
+        switch (o) {
+            case Point(int x, int y, double z) when (x < 3 && y != 6) -> {}
+             // 2 violations above:
+            //                  ''3' is a magic number.'
+            //                   ''6' is a magic number.'
+            case Point(_, _, double z) when z > (10.88) -> {}
+            // violation above, ''10.88' is a magic number'
+            case String s when s.length() != 6 -> {}
+            // violation above, ''6' is a magic number'
+            default -> {}
+        }
+
+        int w = switch (o) {
+            case Point(int x, int y, double z) when (z == 0.5) -> 5;
+             // 2 violations above:
+            //                  ''0.5' is a magic number.'
+           //                   ''5' is a magic number.'
+            case String s -> {
+                yield 6; // violation, ''6' is a magic number'
+            }
+            default -> 0;
+        };
+    }
+
+    record Point(int x, int y, double z) { }
+}

